### ISSUE TITLE
Problems with field arithmetic

### ISSUE THREAD

**Issue by peterdettman** [armfazh/rfc7748_precomputed#5]
I reviewed the field implementation in  fp25519_x64 and either I'm missing something or both the add_EltFp25519_1w_x64 and sub_EltFp25519_1w_x64 have incomplete carry handling.

e.g. I can't see why this line would never carry, since RAX could hold any value, and this line is (possibly) adding 38 to it: https://github.com/armfazh/rfc7748_precomputed/blob/cde6f4990546cceef30fbfecc5063a2dabfe2f91/src/fp25519_x64.c#L676

I think a similar logic also applies for: https://github.com/armfazh/rfc7748_precomputed/blob/cde6f4990546cceef30fbfecc5063a2dabfe2f91/src/fp25519_x64.c#L725

I don't see any tests specifically for add_/sub_. They are covered e.g. by the "mul/sqr" test, but random inputs with 64-bit limbs are very unlikely to catch carry handling errors, especially since random_EltFp25519_1w_x64 is only generating 255-bit values, while add_/sub_ apparently support 256-bit values (?).

I only briefly checked fp448_x64 for comparison, but I believe add_EltFp448_1w_x64 and sub_EltFp448_1w_x64 also have similar errors, probably even less likely to be detected by random inputs.

**Comment by armfazh** [comment]
With respect to the fp25519 arithmetic this was already solved in [42953](https://github.com/armfazh/rfc7748_precomputed/commit/429530dc22cecb9625380f077f6e4c961e540f18). 
You can verify whether the new code computes correctly the arith operations in fp25519.
Thanks for pointing this bug.

**Comment by armfazh** [comment]
> I only briefly checked fp448_x64 for comparison, but I believe add_EltFp448_1w_x64 and sub_EltFp448_1w_x64 also have similar errors, probably even less likely to be detected by random inputs.

More tests are needed for the case of fp448.

**Comment by peterdettman** [comment]
A simple change that will reveal more failures is e.g.:
```
void random_EltFp25519_1w_x64(uint64_t *A)
{
        random_bytes((uint8_t*)A,SIZE_ELEMENT_BYTES);
        for (int i = 0; i < 4; ++i) {
                A[i] = -(A[i] & 1);    // <- Each limb becomes either all 0 bits or all 1 bits
        }
        A[3] &= ((uint64_t)1<<63)-1;    // <- Also comment this out for more failures
}
```

A similar change will reveal errors for fp448 also.

Assuming that add_EltFp25519_1w_x64 is supposed to work correctly for all 256-bit values (i.e. there are no tighter bounds on expected inputs), then I believe you need yet another carry (*38) from r11 to r8. Presumably needed in sub_ also.

**Comment by armfazh** [comment]
>Assuming that add_EltFp25519_1w_x64 is supposed to work correctly for all 256-bit values (i.e. there are no tighter bounds on expected inputs), then I believe you need yet another carry (*38) from r11 to r8. Presumably needed in sub_ also.

Thanks for pinpointing this issue. In fact, the pull request merged in [c79ca5e6](https://github.com/armfazh/rfc7748_precomputed/commit/c79ca5e619b776dede4594a1d8e763d301b0ac06l) now handles addition and subtraction for any value of 256 bits.

**Comment by armfazh** [comment]
> A similar change will reveal errors for fp448 also.

This part of the library still requires more analysis. I will open an issue for this particular part.

@peterdettman Please, let me know if you found more issues in the arithmetic and DH implementation of X25519.
